src/arm/windows/init_by_logical_sys_info.c

```diff
@@ -0,0 +1,842 @@
+	const uint32_t core_id,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX core_info,
+	struct cpuinfo_core* cores,
+	const struct woa_chip_info *chip_info);
+
+static void store_cache_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX info,
+	struct cpuinfo_cache* current_cache);
+
+void connect_packages_cores_clusters_by_processors(
+	struct cpuinfo_processor* processors,
+	const uint32_t nr_of_processors,
+	struct cpuinfo_package* packages,
+	const uint32_t nr_of_packages,
+	struct cpuinfo_cluster* clusters,
+	struct cpuinfo_core* cores,
+	const uint32_t nr_of_cores,
+	const struct woa_chip_info* chip_info,
+	enum cpuinfo_vendor vendor);
+
+static inline uint32_t low_index_from_kaffinity(KAFFINITY kaffinity);
+
+
+BOOL cpu_info_init_by_logical_sys_info(
+	const struct woa_chip_info *chip_info,
+	const enum cpuinfo_vendor vendor)
+{
+	struct cpuinfo_processor* processors = NULL;
+	struct cpuinfo_package* packages = NULL;
+	struct cpuinfo_cluster* clusters = NULL;
+	struct cpuinfo_core* cores = NULL;
+	struct cpuinfo_cache* caches = NULL;
+	struct cpuinfo_uarch_info* uarchs = NULL;
+
+	uint32_t nr_of_packages = 0;
+	uint32_t nr_of_cores = 0;
+	uint32_t nr_of_all_caches = 0;
+	uint32_t numbers_of_caches[MAX_NR_OF_CACHES] = {0};
+	
+	uint32_t nr_of_uarchs = 0;
+	bool result = false;
+	
+	HANDLE heap = GetProcessHeap();
+
+	/* 1. Count available logical processor groups and processors */
+	const uint32_t max_group_count = (uint32_t) GetMaximumProcessorGroupCount();
+	cpuinfo_log_debug("detected %"PRIu32" processor group(s)", max_group_count);
+	/* We need to store the absolute processor ID offsets for every groups, because
+	 *  1. We can't assume every processor groups include the same number of
+	 *     logical processors.
+	 *  2. Every processor groups know its group number and processor IDs within
+	 *     the group, but not the global processor IDs.
+	 *  3. We need to list every logical processors by global IDs.
+	*/
+	uint32_t* global_proc_index_per_group =
+		(uint32_t*) HeapAlloc(heap, 0, max_group_count * sizeof(uint32_t));
+	if (global_proc_index_per_group == NULL) {
+		cpuinfo_log_error(
+			"failed to allocate %zu bytes for descriptions of %"PRIu32" processor groups",
+			max_group_count * sizeof(struct cpuinfo_processor), max_group_count);
+		goto clean_up;
+	}
+	
+	uint32_t nr_of_processors =
+		count_logical_processors(max_group_count, global_proc_index_per_group);
+	processors = HeapAlloc(heap, HEAP_ZERO_MEMORY, nr_of_processors * sizeof(struct cpuinfo_processor));
+	if (processors == NULL) {
+		cpuinfo_log_error(
+			"failed to allocate %zu bytes for descriptions of %"PRIu32" logical processors",
+			nr_of_processors * sizeof(struct cpuinfo_processor), nr_of_processors);
+		goto clean_up;
+	}
+
+	/* 2. Read topology information via MSDN API: packages, cores and caches*/
+	nr_of_packages = read_packages_for_processors(

```

Comment:
there are many things that can go wrong in these calls, but there is no way to chain the error and exit the function on error. That's not great for defensive code.
Consider adding throw/catch if permitted.

Reply:
As far as I see, there are not throw/catch in the current codebase and I'd think that's intentional, so I'd try to cleanup the error paths and make sure it's going to be safe.

Reply:
Yes `C` code mostly